### PR TITLE
Add more reinforced blocks

### PR DESCRIPTION
**Pull request by RedstoneDubstep**
In this PR I've added all the Reinforced Block I mentioned earlier on the server (+ the Reinforced Grass Block), and I've done some minor changes to the blockstate files of some existing Reinforced Blocks to let them look more natural and more like their Vanilla counterpart.
I haven't found the time to update the changelog, it would be cool if one of you guys could do that.

### CODE REVIEW

File: src/main/resources/assets/securitycraft/models/item/reinforced_cobweb.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "parent": "item/generated",
+    "textures": {
+        "layer0": "securitycraft:item/reinforced_cobweb"
```
Comment (bl4ckscor3) [inline src/main/resources/assets/securitycraft/models/item/reinforced_cobweb.json:4]: Use the vanilla cobweb texture here (`block/cobweb`) and remove the one you added. The item tint will be applied nonetheless.

File: src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java [hunk 2]
```diff
@@ -264,6 +270,9 @@ public void tint()
 		toTint.put(SCContent.CRYSTAL_QUARTZ_SLAB.get(), crystalQuartzTint);
 		toTint.put(SCContent.STAIRS_CRYSTAL_QUARTZ.get(), crystalQuartzTint);
 		toTint.forEach((block, tint) -> Minecraft.getInstance().getBlockColors().register((state, world, pos, tintIndex) -> {
+			if (block == SCContent.REINFORCED_GRASS_BLOCK.get())
+				if (!world.getBlockState(pos).get(ReinforcedSnowyDirtBlock.SNOWY))
+					return world != null && pos != null ? BiomeColors.getGrassColor(world, pos) : GrassColors.get(0.5D, 1.0D);
 			if(tintBlocks)
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java:276]:
```suggestion
			if(block == SCContent.REINFORCED_GRASS_BLOCK.get() && !world.getBlockState(pos).get(ReinforcedSnowyDirtBlock.SNOWY))
			{
				if(tintIndex == 0)
					return tintBlocks ? tint : noTint;

				int grassTint = world != null && pos != null ? BiomeColors.getGrassColor(world, pos) : GrassColors.get(0.5D, 1.0D);

				return tintBlocks ? mixTints(grassTint, tint) : grassTint;
			}
			else if(tintBlocks)
```

File: src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java [hunk 3]
```diff
@@ -272,6 +281,8 @@ else if(tint == reinforcedCrystalQuartzTint || tint == crystalQuartzTint)
 				return noTint;
 		}, block));
 		toTint.forEach((item, tint) -> Minecraft.getInstance().getItemColors().register((stack, tintIndex) -> {
+			if (item == SCContent.REINFORCED_GRASS_BLOCK.get())
+				return GrassColors.get(0.5D, 1.0D);
 			if(tintBlocks)
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java:286]:
```suggestion
			if(item == SCContent.REINFORCED_GRASS_BLOCK.get())
			{
				if(tintIndex == 0)
					return tintBlocks ? tint : noTint;

				int grassTint = GrassColors.get(0.5D, 1.0D);
				return tintBlocks ? mixTints(grassTint, tint) : grassTint;
			}
			else if(tintBlocks)
```

File: src/main/resources/assets/securitycraft/models/block/reinforced_grass_block.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            "to": [ 16, 16, 16 ],
+            "faces": {
+                "north": { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 0, "cullface": "north" },
+                "south": { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 0, "cullface": "south" },
+                "west":  { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 0, "cullface": "west" },
+                "east":  { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 0, "cullface": "east" }
+            }
+        }
+    ]
+}
```
Comment (bl4ckscor3) [inline src/main/resources/assets/securitycraft/models/block/reinforced_grass_block.json:32]:
```suggestion
{
    "parent": "block/block",
    "textures": {
        "overlay": "block/grass_block_side_overlay",
        "particle": "block/dirt",
        "bottom": "block/dirt",
        "top": "block/grass_block_top",
        "side": "block/grass_block_side"
    },
    "elements": [
        {   "from": [ 0, 0, 0 ],
            "to": [ 16, 16, 16 ],
            "faces": {
                "down":  { "uv": [ 0, 0, 16, 16 ], "texture": "#bottom", "cullface": "down", "tintindex": 0 },
                "up":    { "uv": [ 0, 0, 16, 16 ], "texture": "#top",    "cullface": "up", "tintindex": 1 },
                "north": { "uv": [ 0, 0, 16, 16 ], "texture": "#side",   "cullface": "north", "tintindex": 0 },
                "south": { "uv": [ 0, 0, 16, 16 ], "texture": "#side",   "cullface": "south", "tintindex": 0 },
                "west":  { "uv": [ 0, 0, 16, 16 ], "texture": "#side",   "cullface": "west", "tintindex": 0 },
                "east":  { "uv": [ 0, 0, 16, 16 ], "texture": "#side",   "cullface": "east", "tintindex": 0 }
            }
        },
        {   "from": [ 0, 0, 0 ],
            "to": [ 16, 16, 16 ],
            "faces": {
                "north": { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 1, "cullface": "north" },
                "south": { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 1, "cullface": "south" },
                "west":  { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 1, "cullface": "west" },
                "east":  { "uv": [ 0, 0, 16, 16 ], "texture": "#overlay", "tintindex": 1, "cullface": "east" }
            }
        }
    ]
}
```
And please remove the custom textures you made for this block.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java [hunk 5]
```diff
@@ -60,6 +90,29 @@ public BaseReinforcedBlock(SoundType soundType, Material mat, Supplier<Block> vB
...
+				TileEntity te = world.getTileEntity(pos);
+
+				if(te instanceof OwnableTileEntity)
+				{
+					OwnableTileEntity ownableTe = (OwnableTileEntity) te;
+
+					if(ownableTe.getOwner().isOwner((PlayerEntity)entity))
+						return;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (RedstoneDubstep) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:107]: What about the entity.setMotionMultiplier? Should I just delete it or move it to another position? Or should I do nothing about it, and this block was just to move the brackets onto the next lines?

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java [hunk 6]
```diff
@@ -72,6 +125,31 @@ public boolean isBeaconBase(BlockState state, IWorldReader world, BlockPos pos,
...
+		net.minecraftforge.common.PlantType type = plantable.getPlantType(world, pos.offset(facing));
+
+		if (plant.getBlock() == Blocks.CACTUS)
+			return this.getBlock() == SCContent.REINFORCED_SAND.get() || this.getBlock() == SCContent.REINFORCED_RED_SAND.get();
+		
+		switch (type) {
+			case Desert: return this.getBlock() == SCContent.REINFORCED_SAND.get() || this.getBlock() == SCContent.REINFORCED_RED_SAND.get();
+			case Cave:   return Block.hasSolidSide(state, world, pos, Direction.UP);
+			case Plains: return SCTags.Blocks.REINFORCED_DIRT.contains(this.getBlock());
+			case Water:  return this.getBlock() == SCContent.FAKE_WATER_BLOCK.get() || this.getFluidState(state) == SCContent.FAKE_WATER.get() ;
```
[the remark this answers is not in the document]
Comment (RedstoneDubstep) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:141]: Now you can place lily pads onto fake water. I thought it is an improvement (and I left it from copying the whole canSustainPlant method from the Block class), but if you want it to be removed, just tell me.

File: src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java [hunk 7]
```diff
@@ -161,7 +164,10 @@ public static void onFMLClientSetup(FMLClientSetupEvent event)
 		RenderTypeLookup.setRenderLayer(SCContent.REINFORCED_DOOR.get(), RenderType.getCutout());
 		RenderTypeLookup.setRenderLayer(SCContent.REINFORCED_GLASS.get(), RenderType.getCutout());
 		RenderTypeLookup.setRenderLayer(SCContent.REINFORCED_GLASS_PANE.get(), RenderType.getCutoutMipped());
+		RenderTypeLookup.setRenderLayer(SCContent.REINFORCED_GRASS_BLOCK.get(), RenderType.getCutout());
```
[the remark this answers is not in the document]
Comment (RedstoneDubstep) [reply src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java:167]: I'll do, but would you like to explain the difference between them, so I understand in the future?

File: src/main/resources/assets/securitycraft/models/item/reinforced_cobweb.json (hunk 1 above)
Comment (RedstoneDubstep) [reply src/main/resources/assets/securitycraft/models/item/reinforced_cobweb.json:4]: Oh,  I didn't know that.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java [hunk 8]
```diff
@@ -60,6 +90,29 @@ public BaseReinforcedBlock(SoundType soundType, Material mat, Supplier<Block> vB
...
+	@Override
+	public void onEntityCollision(BlockState state, World world, BlockPos pos, Entity entity) {
+		if (this == SCContent.REINFORCED_COBWEB.get())
+			if (entity instanceof PlayerEntity)
+			{
+				TileEntity te = world.getTileEntity(pos);
+
+				if(te instanceof OwnableTileEntity)
+				{
+					OwnableTileEntity ownableTe = (OwnableTileEntity) te;
```
[the remark this answers is not in the document]
Comment (RedstoneDubstep) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:102]: I'm sorry for my inconsistent knowledge in this field, but how should I do that? I know there's some easy way but I don't know it, and you can also just tell me where this already happens, so I'll copy from that.

**Comment by RedstoneDubstep** [comment]
...and I messed up again with the commits. Just tell me what I should do to fix it please, I'm to tired to think of that myself now.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (bl4ckscor3) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:104]: Well, you're only using this variable once somewhere else. Just replace that occurance with what you assign to this variable. I understand I have probably worded my comment incorrectly. I meant that casting _here_ is not necessary. Instead, just write `((OwnableTileEntity)te)` in the if statement below instead of `ownableTe`.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java [hunk 9]
```diff
@@ -93,19 +94,21 @@ public BaseReinforcedBlock(SoundType soundType, Material mat, Supplier<Block> vB
...
 			{
 				TileEntity te = world.getTileEntity(pos);
 
-				if(te instanceof OwnableTileEntity)
+				if (te instanceof OwnableTileEntity)
 				{
 					OwnableTileEntity ownableTe = (OwnableTileEntity) te;
 
-					if(ownableTe.getOwner().isOwner((PlayerEntity)entity))
+					if (ownableTe.getOwner().isOwner((PlayerEntity) entity))
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java]: Unnecessary change.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bl4ckscor3) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java]: The logic hasn't changed. The {} are just there to make the code easier to read in this case. Leaving them out changes nothing.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java [hunk 10]
```diff
@@ -93,19 +94,21 @@ public BaseReinforcedBlock(SoundType soundType, Material mat, Supplier<Block> vB
...
 						return;
 				}
 			}
-		entity.setMotionMultiplier(state, new Vec3d(0.25D, (double) 0.05F, 0.25D));
+			entity.setMotionMultiplier(state, new Vec3d(0.25D, (double) 0.05F, 0.25D));
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:110]: New line before this.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bl4ckscor3) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:144]: Yes, please.

File: src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java [hunk 11]
```diff
@@ -237,6 +237,19 @@ public static void onFMLClientSetup(FMLClientSetupEvent event)
...
+		int green = (tint1 >> 0x8) & 0xFF;
+		int blue = tint1 & 0xFF;
+
+		red *= (float)(tint2 >> 0x10 & 0xFF) / 0xFF;
+		green *= (float)(tint2 >> 0x8 & 0xFF) / 0xFF;
+		blue *= (float)(tint2 & 0xFF) / 0xFF;
+
+		return ((red << 8) + green << 8) + blue;
+	}
+
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/network/ClientProxy.java:252]: Move this below the tint method, please. It doesn't really matter for functionality, but it makes more logical sense to me.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (RedstoneDubstep) [reply src/main/java/net/geforcemods/securitycraft/blocks/reinforced/BaseReinforcedBlock.java:104]: Ok I was stupid, I previously tried with `(OwnableTileEntity)te.getOwner()` and it sent me an error because it didn't find `getOwner`, as I missed the brackets around the `(OwnableTileEntity)te`.

File: src/main/java/net/geforcemods/securitycraft/blocks/reinforced/ReinforcedSnowyDirtBlock.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+import net.minecraftforge.common.IPlantable;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.common.PlantType;
+
+import java.util.List;
+import java.util.Random;
+
+public class ReinforcedSnowyDirtBlock extends SnowyDirtBlock implements IReinforcedBlock, IGrowable
+{
+	private Block BLOCK;
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/blocks/reinforced/ReinforcedSnowyDirtBlock.java:36]: For your information: Only static final fields' identifiers should be all uppercase as per convention. You don't need to change this, as I will go through this file later on when I can find the time and clean it up, as most of this seems to be vanilla code.

File: src/main/java/net/geforcemods/securitycraft/datagen/BlockTagGenerator.java [hunk 13]
```diff
@@ -62,11 +62,7 @@ protected void registerTags()
 				SCContent.REINFORCED_PODZOL.get(),
 				SCContent.REINFORCED_MYCELIUM.get());
 		getBuilder(SCTags.Blocks.REINFORCED_END_STONES).add(
-				SCContent.REINFORCED_END_STONE.get(),
-				SCContent.REINFORCED_END_STONE_BRICKS.get(),
-				SCContent.REINFORCED_END_STONE_BRICK_SLAB.get(),
-				SCContent.REINFORCED_END_STONE_BRICK_STAIRS.get(),
-				SCContent.REINFORCED_END_STONE_BRICK_WALL.get());
+				SCContent.REINFORCED_END_STONE.get());
```
Comment (bl4ckscor3) [inline src/main/java/net/geforcemods/securitycraft/datagen/BlockTagGenerator.java]: When changing things in the datagen package, ALWAYS rerun the `runData` run config.
